sched/semaphore/spinlock.c

```diff
@@ -384,4 +384,176 @@ void spin_clrbit(FAR volatile cpu_set_t *set, unsigned int cpu,
+ * Description:
+ *   Initialize the link linsk of fair spinlock.
+ *   If there is any spinlock node, the node will get lock.
+ *
+ * Input Parameters:
+ *   lock_list - A reference to the spinlock list.
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   None.
+ *
+ ****************************************************************************/
+
+void fair_spinlock_list_init(fair_spinlock_list_t *lock_list)
+{
+  lock_list->lock = SP_UNLOCKED;
+  list_initialize(&lock_list->list);
+}
+
+/****************************************************************************
+ * Name: fair_spinlock_init
+ *
+ * Description:
+ *   Initialize the fair spinlock. Tje lock state is set as SP_LOCK util
+ *   some give lock to holder.
+ *
+ * Input Parameters:
+ *   lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void fair_spinlock_init(fair_spinlock_t *lock)
+{
+  lock->lock = SP_LOCKED;
+  lock->holder = this_task();
+  list_initialize(&lock->node);
+}
+
+/****************************************************************************
+ * Name: fair_spin_lock
+ *
+ * Description:
+ *   Try to lock the spinlock and insert this lock to lock list.
+ *   If list is empty, the caller can get the lock.
+ *   If list is not empty, the lock will be inserted to this list.
+ *
+ * Input Parameters:
+ *   lock_list - A reference to fair lock list and insert this lock to list.
+ *   fair_lock - A reference to the spinlock object to lock.
+ *
+ * Returned Value:
+ *   None
+ *
+ * Assumptions:
+ *   Not running at the interrupt level.
+ *
+ ****************************************************************************/
+
+void fair_spin_lock(FAR fair_spinlock_list_t *lock_list,
+                    FAR fair_spinlock_t *fair_lock)
+{
+  FAR struct tcb_s *rtcb = this_task();
+  fair_spinlock_t *entry;
+  fair_spinlock_t *temp;
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SPINLOCKS
+  /* Notify that we have the spinlock */
+
+  sched_note_spinlock(this_task(), &fair_lock->lock, NOTE_SPINLOCK_LOCKED);
+#endif
+
+  /* There is no thread  wait this lock */
+
+  spin_lock(&lock_list->lock);
+
+  /* There is no node hold this lock */
+
+  if (list_is_empty(&lock_list->list))
+    {
+      fair_lock->lock = SP_UNLOCKED;
+      list_add_head(&lock_list->list, &fair_lock->node);
+    }
+
+  /* There is a task hold this lock */
+
+  else
+    {
+      /* priority based link list */
+
+      list_for_every_entry_safe(&lock_list->list, entry, temp,
+                              struct fair_spinlock_s, node)
+        {
+          if (entry->holder->sched_priority > rtcb->sched_priority)
+            {
+              break;
+            }
+        }
+
+      list_add_before(&entry->node, &fair_lock->node);
+    }
+
+  spin_unlock(&lock_list->lock);
+
+  while (up_testset(&fair_lock->lock) == SP_LOCKED)
```

Comment:
I don't see where the code can ensure the lock happen in the first in and first out order?

Reply:
Yes. I thank you are right.
This method only avoid the same thread get lock continuously.

Reply:
Because RTOS is priority-base, I try to designed a policy.
First we have a queue (lock_list) to main threads which want to get this lock.
If queue is empty, the first one can get this lock and when it finishes task it will check the queue is empty or not.
 If it is empty just exit. If there is a thread in the queue, it will give lock to next one in the queue.
 
 And the queue I hope it's priority base so (here)[https://github.com/TaiJuWu/nuttx/blob/fc2b66b139ba0f67706abc7ba88e47c78e692e12/sched/semaphore/spinlock.c#L487] check the priority and put it in appropriate position.
But lock_list spin lock can let many thread spin so I try to use lock-free skill to avoid this problem.

Reply:
yes it is a modified approach to `ticket lock` , that's for optimizing the cache-line, avoiding to invalidate on all cpus, so the `ticket` is not anymore shared but passed by.

Reply:
If you want to use MCS algo to reduce the cache contention, the per CPU structure is required:
https://lwn.net/Articles/590243/
But since NuttX's SMP is designed with a small number of CPU cores(2~4), tick spinlock is enough here. If NuttX want to support 10+ or even 100+ cores, many design must be changed, for example:

1. One big critical section need be removed.
2. One timer interrupt to drive all core scheduling need become per CPU.

and more. Before the core design get fixed, I don't see the benefit to use MCS lock or Qspinlocks.

Reply:
Thank you for your reply.
BTW, how many tasks run on nuttx usually.
Because I found [sched_mergeprioritized.c](https://github.com/apache/nuttx/blob/master/sched/sched/sched_mergeprioritized.c) merge tcb one by one.

Maybe we can merge with chuck as new policy?

Reply:
> Thank you for your reply. BTW, how many tasks run on nuttx usually.

there is no limitation, we have products with run more than 100 tasks/threads.

> Because I found [sched_mergeprioritized.c](https://github.com/apache/nuttx/blob/master/sched/sched/sched_mergeprioritized.c) merge tcb one by one.
> 
> Maybe we can merge with chuck as new policy?

But the spinlock contention doesn't couple with the thread number, but the core number.

Reply:
> > Thank you for your reply. BTW, how many tasks run on nuttx usually.
> 
> there is no limitation, we have products which has more than 100 tasks/threads.
> 
> > Because I found [sched_mergeprioritized.c](https://github.com/apache/nuttx/blob/master/sched/sched/sched_mergeprioritized.c) merge tcb one by one.
> > Maybe we can merge with chuck as new policy?
> 
> But the spinlock contention isn't related to the thread number, but the core number.

Sorry this is another problem, I should discussed here.